src/tracing/agent.cc

```diff
@@ -70,5 +55,100 @@ void Agent::ThreadCb(void* arg) {
   uv_run(&agent->tracing_loop_, UV_RUN_DEFAULT);
 }
 
+void Agent::Enable(const std::string& categories) {
+  if (!categories.empty()) {
+    std::stringstream category_list(categories);
+    while (category_list.good()) {
+      std::string category;
+      getline(category_list, category, ',');
+      categories_.insert(category.c_str());
+    }
+    RestartTracing();
+  }
+}
+
+void Agent::Enable(CategorySet* set) {
+  auto source = set->GetCategories();
+  if (!source.empty()) {
+    categories_.insert(source.begin(), source.end());
+    RestartTracing();
+  }
+}
+
+void Agent::Disable(CategorySet* set) {
+  auto source = set->GetCategories();
+  if (!source.empty()) {
+    for (auto category : source) {
+      auto pos = categories_.lower_bound(category);
+      if (pos != categories_.end())
+        categories_.erase(pos);
+    }
+    RestartTracing();
+  }
+}
+
+void Agent::RestartTracing() {
+  static bool warned;
+  if (!warned) {
+    warned = true;
+    fprintf(stderr, "Warning: Trace event is an experimental feature "
+            "and could change at any time.\n");
+  }
+  Start();  // Start the agent if it hasn't already been started
+  tracing_controller_->StopTracing();
+  auto config = CreateTraceConfig();
+  if (config != nullptr)
+    tracing_controller_->StartTracing(config);
+}
+
+TraceConfig* Agent::CreateTraceConfig() {
+  if (categories_.empty())
+    return nullptr;
+  TraceConfig* trace_config = new TraceConfig();
+  for (auto category = categories_.begin();
+       category != categories_.end();
+       category = categories_.upper_bound(*category)) {
+    trace_config->AddIncludedCategory(category->c_str());
+  }
+  return trace_config;
+}
+
+std::string Agent::GetEnabledCategories() {
+  std::string categories;
+  for (auto category = categories_.begin();
+       category != categories_.end();
+       category = categories_.upper_bound(*category)) {
+    if (!categories.empty())
+      categories += ',';
+    categories += *category;
+  }
+  return categories;
+}
+
+CategorySet::CategorySet(
+    Agent* agent,
+    const std::set<std::string>& categories)
+        : categories_(categories),
+          agent_(agent) {}
+
+CategorySet::~CategorySet() {
+  // Disable if we haven't done so already
+  Disable();
+}
+
+void CategorySet::Enable() {
+  if (!enabled_ && agent_ != nullptr) {
+    agent_->Enable(this);
+    enabled_ = true;
+  }
+}
+
+void CategorySet::Disable() {
+  if (enabled_ && agent_ != nullptr) {
+    agent_->Disable(this);
+    enabled_ = false;
+  }
+}
```

Comment:
These methods look unnecessary.  I can't really tell why `CategorySet` is exposed to JS because the core of its logic is just this and that could be done by a single binding method:
```c++
Utf8Value categories(args.GetIsolate(), args[0]);
v8_platform.GetTracingAgent()->Enable(categories);
```
And `Enable()` can be condensed to just this, it looks like:
```c++
// split categories by comma, then:
TraceConfig* trace_config = new TraceConfig();
for (auto category : categories) trace_config->AddIncludedCategory(category.c_str());
tracing_controller_->StartTracing(config);
```
If `CategorySet` exists to have some weak GC object, eh, there's probably better ways to achieve that.  I see it's been discussed at some length already. :-)

Reply:
The reason `CategorySet` exists is specifically to have some weak GC object that can be used with the theoretical goal of being able to manage multiple trace event targets and possibly other hypothetical use cases later. I'm personally not convinced it's needed and would prefer the `trace_events.enable()`/ `trace_events.disable()` approach with the internal ref counting. It would make this much simpler as you suggest.

Reply:
To explain a bit more... the simplified versions you suggest are definitely possible. However,

1) there is a general concern over multiple modules stepping on each other if they are enabling/disabling categories which is why the code uses a crude form of reference counting using an std::multiset as opposed to using `AddIncludedCategory` directly. We should likely keep that in place.

2) `CategorySet` exists specifically to have a JS that can be used. Right now enabling and disabling the categories is the only use for it. There has been some discussion, however, about allowing JS code to specify multiple trace event destinations (e.g. one `Tracing` object could direct the trace events it cares about to one file, while another directs trace events it wants to another file). This is a purely hypothetical use case at this point but it's been discussed.

I've got zero problems with simplifying this by eliminating `CategorySet` but we need to resolve the API question... do we, or we do not, want the JS-accessible `Tracing` object?

Reply:
Reading the comment again I realize I misread your original point. The internal `CategorySet` is there largely to eventually allow for more `Tracing` object specific configuration stuff such as multiple outputs. You're absolutely right that it can be eliminated if all we're going to do is enable/disable categories.